network/components/mux_tx.c

```diff
@@ -1,227 +1,146 @@
-{
-    if (addr >= shared_dma_vaddr_cli0 && addr < shared_dma_vaddr_cli0 + DMA_SIZE) {
-        return CLIENT_0;
-    } else if (addr >= shared_dma_vaddr_cli1 && addr < shared_dma_vaddr_cli1 + DMA_SIZE) {
-        return CLIENT_1;
-    }else if (addr >= shared_dma_vaddr_arp && addr < shared_dma_vaddr_arp + DMA_SIZE) {
-        return ARP;
+int extract_offset(uintptr_t *phys) {
+    for (int client = 0; client < NUM_CLIENTS; client++) {
+        if (*phys >= state.buffer_region_paddrs[client] && 
+            *phys < state.buffer_region_paddrs[client] + state.tx_ring_clients[client].free_ring->size * BUFF_SIZE) {
+            *phys = *phys - state.buffer_region_paddrs[client];
+            return client;
+        }
     }
-    print("MUX TX|ERROR: Buffer out of range\n");
-    assert(0);
-    return NUM_CLIENTS;
+    return -1;
 }
 
-/*
- * Loop over all used tx buffers in client queues and enqueue to driver.
- */
-void process_tx_ready(void)
+void tx_provide(void)
 {
-    bool enqueued = 0;
-    int err;
-    process_tx_ready_:
+    bool enqueued = false;
     for (int client = 0; client < NUM_CLIENTS; client++) {
-        while (!ring_empty(state.tx_ring_clients[client].used_ring) && !ring_full(state.tx_ring_drv.used_ring)) {
-            uintptr_t addr;
-            unsigned int len;
-            void *cookie;
-            uintptr_t phys;
-
-            err = dequeue_used(&state.tx_ring_clients[client], &addr, &len, &cookie);
-            assert(!err);
-            phys = get_phys_addr(addr);
-            assert(phys);
-            err = enqueue_used(&state.tx_ring_drv, phys, len, cookie);
-            assert(!err);
-
-            enqueued = true;
-        }
-
-        state.tx_ring_clients[client].used_ring->notify_reader = true;
-
-        THREAD_MEMORY_FENCE();
-
-        if (!ring_empty(state.tx_ring_clients[client].used_ring) && !ring_full(state.tx_ring_drv.used_ring)) {
-            state.tx_ring_clients[client].used_ring->notify_reader = false;
-            goto process_tx_ready_;
+        bool reprocess = true;
+        while (reprocess) {
+            while (!ring_empty(state.tx_ring_clients[client].used_ring)) {
+                buff_desc_t buffer;
+                int err = dequeue_used(&state.tx_ring_clients[client], &buffer);
+                assert(!err);
+
+                if (buffer.phys_or_offset % BUFF_SIZE || 
+                    buffer.phys_or_offset >= BUFF_SIZE * state.tx_ring_clients[client].used_ring->size) {
+                    dprintf("MUX_TX|LOG: Client provided offset %llx which is not buffer aligned or outside of buffer region\n", buffer.phys_or_offset);
+                    err = enqueue_free(&state.tx_ring_clients[client], buffer);
+                    assert(!err);
+                    continue;
+                }
+
+                microkit_arm_vspace_data_clean(buffer.phys_or_offset + state.buffer_region_vaddrs[client], buffer.phys_or_offset + state.buffer_region_vaddrs[client] + buffer.len);
```

Comment:
Sorry I should have been clearer with my comment about the `microkit_arm_vspace_*` API. These are just wrapping over the `seL4_ARM_VSpace_*` operations so as the user you don't have to deal with the VSpace cap.

I think here we would want to keep doing the user-space cache operations so `cache_clean` as you had it before.